presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/CachingHiveMetastore.java

```diff
@@ -535,11 +591,39 @@ protected void invalidateTable(String databaseName, String tableName)
             String tableName,
             Map<Column, Domain> partitionPredicates)
     {
+        if (partitionVersioningEnabled) {
+            List<PartitionNameWithVersion> partitionNamesWithVersion = getPartitionNamesWithVersionByFilter(databaseName, tableName, partitionPredicates);
+            List<String> result = partitionNamesWithVersion.stream().map(PartitionNameWithVersion::getPartitionName).collect(toImmutableList());
+            partitionNamesWithVersion.forEach(partitionNameWithVersion -> invalidateStalePartition(partitionNameWithVersion, databaseName, tableName));
+            return result;
+        }
         return get(
                 partitionFilterCache,
                 partitionFilter(databaseName, tableName, partitionPredicates));
     }
 
+    @Override
+    public List<PartitionNameWithVersion> getPartitionNamesWithVersionByFilter(
+            String databaseName,
+            String tableName,
+            Map<Column, Domain> partitionPredicates)
+    {
+        return delegate.getPartitionNamesWithVersionByFilter(databaseName, tableName, partitionPredicates);
+    }
+
+    private void invalidateStalePartition(PartitionNameWithVersion partitionNameWithVersion, String databaseName, String tableName)
+    {
+        HivePartitionName hivePartitionName = hivePartitionName(databaseName, tableName, partitionNameWithVersion.getPartitionName());
+        Optional<Partition> partition = partitionCache.getIfPresent(hivePartitionName);
+        if (partition != null && partition.isPresent()) {
+            Optional<Integer> partitionVersion = partition.get().getPartitionVersion();
+            if (partitionVersion.isPresent() && partitionVersion.get() < partitionNameWithVersion.getPartitionVersion()) {
```

Comment:
This function is only called when `partitionVersioningEnabled` is `true`, so I'm wondering if we should actually treat version not existing as stale? I don't have preference, just a random thought.

Reply:
Prior to this PR, there might be users who are enabling caching by the setting the params `hive.metastore-cache-ttl` etc. Now when they upgrade to this PR , and we consider partitions with NO version  as stale and delete them, then we might break the caching feature for them. WDYT ?

Reply:
I made changes to consider the partition stale if the version is empty. This will also help testing.